streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -605,9 +605,9 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
     f: OutElem => ZIO[Env1, OutErr1, OutElem2]
   )(implicit trace: Trace): ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] = {
     lazy val reader: ZChannel[Env1, OutErr, OutElem, OutDone, OutErr1, OutElem2, OutDone] =
-      ZChannel.readWith(
+      ZChannel.readWithCause(
         (out: OutElem) => ZChannel.fromZIO(f(out)).flatMap(ZChannel.write(_)) *> reader,
-        (e: OutErr1) => ZChannel.fail(e),
+        (e: Cause[OutErr1]) => ZChannel.failCause(e),
```

Comment:
Isn't `readWith` doing the same automatically?
https://github.com/zio/zio/blob/series/2.x/streams/shared/src/main/scala/zio/stream/ZChannel.scala#L1803

Reply:
The issue is if we extract an `E` in `failureOrCause` and the fail with that `E` we lose the tracing information associated with the original cause.

Reply:
Strikes me we should probably have `ZChannel.refailCause` at some point.

Reply:
💯